### PR TITLE
Add a timeout to the sshDialer to prevent indefinite hangs.

### PR DESCRIPTION
**Pull request by cjcullen**
Prevents the SSH Dialer from hanging forever. Fixes a problem where SSH Tunnels get stuck trying to open.

Addresses #23835.

### CODE REVIEW

**Comment by cjcullen** [comment]
@a-robinson @roberthbailey

File: pkg/ssh/ssh.go [hunk 1]
```diff
@@ -163,11 +163,38 @@ func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*s
 	return ssh.Dial(network, addr, config)
 }
 
+// timeoutDialer wraps an sshDialer with a timeout around Dial().
```
Comment (a-robinson) [inline pkg/ssh/ssh.go]: Can you add a comment either here or up above where it's used with a little more context on why this is needed?

File: pkg/ssh/ssh.go [hunk 2]
```diff
@@ -163,11 +163,38 @@ func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*s
...
+type timeoutDialer struct {
+	dialer  sshDialer
+	timeout time.Duration
+}
+
+const sshDialTimeout = 150 * time.Second
```
Comment (a-robinson) [inline pkg/ssh/ssh.go]: Leave a comment on why we chose 150 seconds

File: pkg/ssh/ssh.go [hunk 3]
```diff
@@ -163,11 +163,38 @@ func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*s
...
+		defer runtime.HandleCrash()
+		var err error
+		client, err = d.dialer.Dial(network, addr, config)
+		errCh <- err
+	}()
+	select {
+	case err := <-errCh:
+		return client, err
+	case <-time.After(d.timeout):
+		return nil, fmt.Errorf("Timed out dialing %s:%s", network, addr)
```
Comment (a-robinson) [inline pkg/ssh/ssh.go]: nit: use lower case for error messages

File: pkg/ssh/ssh.go [hunk 4]
```diff
@@ -163,11 +163,38 @@ func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*s
...
+
+var realTimeoutDialer sshDialer = &timeoutDialer{&realSSHDialer{}, sshDialTimeout}
+
+func (d *timeoutDialer) Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
+	var client *ssh.Client
+	errCh := make(chan error, 1)
+	go func() {
+		defer runtime.HandleCrash()
+		var err error
+		client, err = d.dialer.Dial(network, addr, config)
```
Comment (a-robinson) [inline pkg/ssh/ssh.go]: I'm not super familiar with go's barrier model for order of execution across a channel send, but I'd guess it's safe to assume that the client variable gets set before err gets sent on the channel? The alternative would be to pass both the client and err via the channel.

File: pkg/ssh/ssh_test.go [hunk 5]
```diff
@@ -328,3 +328,39 @@ func TestSSHUser(t *testing.T) {
...
+	}{
+		// delay > timeout should cause ssh.Dial to timeout.
+		{1 * time.Second, 0, nil, "Timed out dialing"},
+		// delay < timeout should return the result of the call to the dialer.
+		{0, 1 * time.Second, nil, ""},
+		{0, 1 * time.Second, fmt.Errorf("Test dial error"), "Test dial error"},
+	}
+	for _, tc := range testCases {
+		dialer := &timeoutDialer{&slowDialer{tc.delay, tc.err}, tc.timeout}
+		_, err := dialer.Dial("tcp", "127.0.0.1:8080", &ssh.ClientConfig{})
```
Comment (a-robinson) [inline pkg/ssh/ssh_test.go]: Will this test fail if it's run on a machine that has something listening on port 8080?

Comment (a-robinson) [inline pkg/ssh/ssh_test.go]: Ah, nevermind, it's never actually dialed.

File: pkg/ssh/ssh.go (hunk 3 above)
Comment (cjcullen) [inline pkg/ssh/ssh.go]: Done.

File: pkg/ssh/ssh.go (hunk 2 above)
Comment (cjcullen) [inline pkg/ssh/ssh.go]: Done.

File: pkg/ssh/ssh.go (hunk 1 above)
Comment (cjcullen) [inline pkg/ssh/ssh.go]: Done.

**Comment by a-robinson** [comment]
LGTM

File: pkg/ssh/ssh.go (hunk 4 above)
Comment (cjcullen) [inline pkg/ssh/ssh.go]: Yeah, it's safe: https://golang.org/ref/mem#tmp_7

**Comment by roberthbailey** [comment]
LGTM.

**Comment by cjcullen** [comment]
@k8s-bot unit test this issue: #21640

**Pull request merged**